### PR TITLE
feat(cli): Add `log-format` cli option

### LINKED ISSUE #1709: JSON format option for Vector's internal logs
This is hopefully an easy option to add, but it would be nice to provide an option to format our internal logs as JSON. The current format is not machine friendly:

```
Feb 04 15:26:54.890 DEBUG sink{name=sink0 type=logdna}: vector::sinks::util::http: response. status=403 Forbidden version=HTTP/1.1
```

Given the following global option:

```toml
[observability]
log_format = "json"
```

or CLI flag:

```
vector --log-format="json"
```

(I'm open to other naming schemes)

Would produce

```json
{"timestamp": "2020-02-04T15:26:54.890Z", "level": "debug", "sink": {"name": "sink0", "type": "logdna"}, "namespace": "vector::sinks::util::http", "message": "response", "status": "403 Forbidden", "version": "HTTP/1.1"}
```

A couple of things to note:

1. The timestamp is in ISO8601 format.
2. The `level` is lowercase.

### PR DESCRIPTION
**Pull request by LucioFranco**
Example json output:

```json
{"timestamp":"Feb 20 11:28:15.096","level":"INFO","target":"vector","fields":{"message":"Log level \"info\" is enabled."}}
{"timestamp":"Feb 20 11:28:15.096","level":"INFO","target":"vector","fields":{"message":"Loading configs.","path":"[\"/etc/vector/vector.toml\"]"}}
{"timestamp":"Feb 20 11:28:15.108","level":"INFO","target":"vector","fields":{"message":"Vector is starting.","version":"0.8.0","git_version":"v0.7.0-168-g841a8f8","released":"Wed, 19 Feb 2020 20:17:41 +0000","arch":"x86_64"}}
{"timestamp":"Feb 20 11:28:15.114","level":"INFO","target":"vector::topology","fields":{"message":"Running healthchecks."}}
{"timestamp":"Feb 20 11:28:15.114","level":"INFO","target":"vector::topology","fields":{"message":"Starting source \"stdin\""}}
{"timestamp":"Feb 20 11:28:15.114","level":"INFO","target":"vector::topology","fields":{"message":"Starting sink \"loki\""}}
{"timestamp":"Feb 20 11:28:15.115","level":"INFO","span":{"name":"source","type":"stdin"},"target":"vector::sources::stdin","fields":{"message":"Capturing STDIN"}}
{"timestamp":"Feb 20 11:28:15.120","level":"ERROR","target":"vector::topology::builder","fields":{"message":"Healthcheck: Failed Reason: error trying to connect: Connection refused (os error 111)"}}
^C{"timestamp":"Feb 20 11:28:17.057","level":"INFO","target":"vector","fields":{"message":"Shutting down."}}
```

Closes #1709

Signed-off-by: Lucio Franco <luciofranco14@gmail.com>

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -131,6 +141,21 @@ impl std::str::FromStr for Color {
     }
 }
 
+impl std::str::FromStr for LogFormat {
```
Comment (bruceg) [inline src/main.rs]: I thought structopt could derive this for arguments, but I guess it only works for sub-commands. Oh well.

File: src/main.rs [hunk 2]
```diff
@@ -195,8 +220,15 @@ fn main() {
 
     let (metrics_controller, metrics_sink) = metrics::build();
 
+    let json = match &opts.log_format {
+        Some(LogFormat::Text) => false,
+        Some(LogFormat::Json) => true,
+        None => false,
+    };
+
```
Comment (bruceg) [inline src/main.rs]: Can't this simply be `let json = opts.log_format == Some(LogFormat::Json);`?
Or, to make the default explicit, `let json = opts.log_format.unwrap_or(LogFormat::Text) == LogFormat::Json;`

File: src/trace.rs [hunk 3]
```diff
@@ -11,17 +11,32 @@ use tracing_subscriber::{layer::SubscriberExt, FmtSubscriber};
...
-pub fn init(color: bool, levels: &str, metrics: Option<metrics::Sink>) {
-    let subscriber = FmtSubscriber::builder()
-        .with_ansi(color)
-        .with_env_filter(levels)
-        .finish()
-        .with(Limit::default());
+pub fn init(color: bool, json: bool, levels: &str, metrics: Option<metrics::Sink>) {
+    let dispatch = if json {
+        let subscriber = FmtSubscriber::builder()
+            .with_ansi(color)
```
Comment (bruceg) [inline src/trace.rs]: Is there a use case for adding color to JSON logs? I thought they would normally be consumed only by other code.

**Comment by binarylogic** [comment]
The ‘fields’ root key is unfortunate. Can that not be flattened?

**Comment by Hoverbear** [comment]
Agreed, I think having `fields` flattened is probably least surprising.

File: src/trace.rs (hunk 3 above)
Comment (LucioFranco) [reply src/trace.rs]: I believe this case most likely gets ignored but I'll remove this :+1:

**Comment by LucioFranco** [comment]
Ok, so I spent some time figuring out if this would even be possible with tracing's current implementation: 

```json
{"timestamp":"Feb 24 14:12:30.265","level":"INFO","target":"vector","message":"Log level \"info\" is enabled."}
{"timestamp":"Feb 24 14:12:30.265","level":"INFO","target":"vector","message":"Loading configs.","path":"[\"/etc/vector/vector.toml\"]"}
{"timestamp":"Feb 24 14:12:30.277","level":"INFO","target":"vector","message":"Vector is starting.","version":"0.8.0","git_version":"v0.7.0-175-gb33cc49","released":"Mon, 24 Feb 2020 18:34:30 +0000","arch":"x86_64"}
{"timestamp":"Feb 24 14:12:30.282","level":"INFO","target":"vector::topology","message":"Running healthchecks."}
{"timestamp":"Feb 24 14:12:30.282","level":"INFO","target":"vector::topology","message":"Starting source \"stdin\""}
{"timestamp":"Feb 24 14:12:30.282","level":"INFO","target":"vector::topology","message":"Starting sink \"loki\""}
{"timestamp":"Feb 24 14:12:30.283","level":"INFO","span":{"name":"source","type":"stdin"},"target":"vector::sources::stdin","message":"Capturing STDIN"}
{"timestamp":"Feb 24 14:12:30.288","level":"ERROR","target":"vector::topology::builder","message":"Healthcheck: Failed Reason: error trying to connect: Connection refused (os error 111)"}
^C{"timestamp":"Feb 24 14:12:39.780","level":"INFO","target":"vector","message":"Shutting down."}
```

This is the output I am getting now, is this along the lines of what you'd like @binarylogic? If so we should allocate time to upstream this into tracing because I don't want to maintain a fork but upstreaming this will probably be somewhat of a contentious topic.

**Comment by binarylogic** [comment]
Yep, but it sounds like there are caveats? Otherwise this looks good.

**Comment by LucioFranco** [comment]
@binarylogic Yeah, I think I'd like to punt on flattened json until we really need it.

**Comment by binarylogic** [comment]
Uh, I would like to flatten the fields. I would never advocate anyone to structure their logging data that way and we’re an observability tool. We should be able to produce simple log messages and follow our own advice. And could you explain why this is hard and what the caveats are? And please revert this until we have consensus.

**Comment by LucioFranco** [comment]
@binarylogic Ok, I've opened a revert PR.

There are a couple reasons I think flattening everything is hard right now:

1) Its not clear what the performance impacts are, if we want to flatten we must support nested spans and possible duplicate keys. To support this we would then need to prefix things like `span.type` etc. Because of this it would actually require a bunch of allocations every time we log which could be very expensive. This isn't measured and if we do implement that we would want to benchmark. I have already seen tons of logging performance issues with naive approaches (in other libraries) and tracing already walks this fine line.

2) There is no consensus yet with the `tracing` folks on what a flattened version would look like. So if we want this feature we would need to invest time into this. Not only would we need to invest time into a correct format that works but also implementing this is non trivial.

To me it seems simpler to ship json as is because I don't think its worth it for us to maintain our own fork of a tracing-subscriber. So to me it either looks like we invest time into getting tracing to output what we want or we don't ship json? In the end this PR seems like a middle ground that is good.

**Comment by Hoverbear** [comment]
@LucioFranco Please open a new PR with this feature we can re-address. :)

**Pull request merged**